### DISCUSSION TITLE
>>Question about Yaw calculation formula in nominalYawGps()

### DISCUSSION THREAD

**Discussion by ZGJ13** [GeoscienceAustralia/ginan#69] [Q&A]
Dear Ginan term
First of all, thank you for providing such a marvelous and comprehensive source code!
I am recently reading on the part of yaw attitude model. When using sun elevation angle namely beta and orbital angle namely mu to calculate yaw angle, you implemented the formula by adding an extra PI in function nominalYawGps(). However, It does't correspond to the methods in most references, and is also inconsistent with the implementation of common open source codes such as groops/rtklib. Do you have any considerations for doing this operation? If  PI is not added to the equation, is the subsequent code also correct? I would appreciate it if you could explain this issue to me. Thank you very much, and I look forward to your reply.
![yaw](https://github.com/GeoscienceAustralia/ginan/assets/66286176/a6c7fb58-0985-4279-af3f-f40e9c7ec956)

**Comment by seballgeyer** [comment]
Hello, 

After investigating the code in question,  the difference in how we calculate the yaw angle in our code compared to other open-source codes like groops/rtklib comes down to how we define the satellite's body reference frame. In our code, we assume that the -Z axis points toward the antenna/Earth, while in other codes, the convention is that the +Z axis points toward the antenna.

This difference affects how we calculate the yaw angle using the sun elevation angle and orbital angle. By adding an extra PI in our calculation, we adjust for this difference in reference frames. If we didn't make this adjustment, it could lead to inconsistencies in the yaw angle calculation and potentially affect the satellite's attitude, which could, in turn, impact the results obtained from the code, as the attitude is used for the PCO, PCV and phase windup.

If you have any further questions or concerns, please don't hesitate to reach out for clarification or additional information.

**Reply by ZGJ13** [reply]
Hello,
Sorry for late response. Thanks a lot for your detailed answer! I've understood the importance of keeping the Z-axis direction consistent in your code.
I don’t know enough about satellite orbit and celestial reference frame, so I still can’t understand is why the defined direction of the Z-axis needs to be reversed. ”-Z axis points toward the antenna/Earth“ is different from the convention in IGS or the definition in Montenbruck et al. (2015), could you please explain the reason to me in detail?
Besides, as for Galileo-2 FOC yaw attitude in function satYawGalFoc(), I find you also call the nominalYawGps() to calculate the nominal yaw, which is inconsistent with Galileo's official documentation(https://www.gsc-europa.eu/support-to-developers/galileo-satellite-metadata#3.1.2). Are these two methods equivalent? Even if they are equivalent. Since the yaw attitude calculation formula in the official document has nothing to do with the sun elevation angle(beta), even when beta is small, there is no need to consider the change in sign of the beta angle during the maneuvering period. Would it be more reasonable to adopt the official nominal calculation method?
Thank you again and look forward for your reply!
![image](https://github.com/GeoscienceAustralia/ginan/assets/66286176/53b62aae-1958-4b57-ae44-c57ce9128cdf)
